File: Python3/169. Majority Element.py
class Solution:
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # numsDict = {}
        # threshold = math.floor(len(nums)/2)
        # for num in nums:
        #     if num not in numsDict:
        #         numsDict[num] = 1
        #     else:
        #         numsDict[num] += 1
        # for num in numsDict:
        #     if numsDict[num] > threshold:
        #         return num
        
        # 这里本来想用排序: return sorted(nums)[len(nums)//2] 
        # 但是后来发现排序这个是不对的，比如[3, 3, 3, 4, 5, 6]
        
        numsDict = collections.Counter(nums) # 这个用的也是映射，不过是collections包
        return max(numsDict.keys(), key=numsDict.get)

File: Golang/169. Majority Element.go
// func majorityElement(nums []int) int { // 32ms
//     threshold := len(nums) / 2
//     numsDict := make(map[int]int, threshold+1)
//     for _, num := range nums {
//         numsDict[num]++
//     }
//     ans := 0
//     for num := range  numsDict {
//         if numsDict[num] > threshold {
//             ans = num
//             break
//         }
//     }
//     return ans
// }

func majorityElement(nums []int) int { // 尝试一下新的写法 // 32ms
    threshold := len(nums) / 2
    numsDict := make(map[int]int, threshold+1)
    for _, num := range nums {
        if numsDict[num]++; numsDict[num] > threshold {
            return num
        }
    }
    return 9999
}
